### ISSUE TITLE
Investigate how to measure memory usage in each external C module

### ISSUE THREAD

**Issue by pmnoxx** [near/nearcore#3430]
Previously I added a measurement of memory usage in `neard`.
I noticed that in a day the memory usage grew from 1.2gb to 3.9gb. That's 2.7gb increase.
Meanwhile allocated memory by Rust code grew from 400mb to 1.4gb (1.7gb mapped).
main thread:  872151 allocations 593 mb  (up from 25275 / 49mb)
peermanager: 37890 allocations 275 mb (up from 8566/28mb)
33 peer actors: 55 allocations each, total 503mb. (up from 29 peer actors / 55 allocations / 422mb) 
other threads use less than 20mb

Meanwhile RocksDB usage went up from 50 mb to 322mb.

I can measure 1.25gb memory usage from Rust and RocksDB report. I'm going to investigate on how to measure memory usage for C code to find out which library is responsible for the unaccounted 1.35gb increase .

**Comment by pmnoxx** [comment]
https://stackoverflow.com/questions/262439/create-a-wrapper-function-for-malloc-and-free-in-chttps://stackoverflow.com/questions/6083337/overriding-malloc-using-the-ld-preload-mechanism

**Comment by pmnoxx** [comment]
The next step is to investigate how rust to C linking works, and  then write a C proxy alloc module, which can measure the amount of memory allocated/freed.

**Comment by pmnoxx** [comment]
I created a proxy allocator for C code, it allows me to analyze memory leaks.

It's able to print stack traces when big memory allocations/dealocations/mmaps get exectued. This allows me, to match given memory region to C/Rust code which allocated it.
Currently, I implemented logging allocations to files and counting the number of currently allocated objects per thread.

In addition I implemented adding headers to each memory allocation. Currently it's unused, but I'm planning to use to for the following.
1) add magic value, this is useful for sanity checks. This will allow me to mark, beginning of allocation, which is useful when analyzing memory dumps.
2) add allocation size to the header, this will allow me to count total size of allocation per thread, and even per function
3) thread id
4) add function pointer to the header, this will allow me to see which function allocated given region.

Given, the data above, I'll be able to generate statistics based on memory dump. It will show us, which library, which function, which thread is leaking memory.
